torchdata/datapipes/iter/load/fsspec.py

```diff
@@ -49,37 +49,55 @@ class FSSpecFileListerIterDataPipe(IterDataPipe[str]):
 
     def __init__(
         self,
-        root: str,
+        root: Union[str, Sequence[str], IterDataPipe],
         masks: Union[str, List[str]] = "",
     ) -> None:
         _assert_fsspec()
 
-        self.root: str = root
+        if isinstance(root, str):
+            root = [
+                root,
+            ]
+        if not isinstance(root, IterDataPipe):
+            self.datapipe: Union[IterDataPipe, IterableWrapper] = IterableWrapper(root)
+        else:
+            self.datapipe = root
         self.masks = masks
 
     def __iter__(self) -> Iterator[str]:
-        fs, path = fsspec.core.url_to_fs(self.root)
-        is_local = fs.protocol == "file" or not self.root.startswith(fs.protocol)
-        if fs.isfile(path):
-            yield self.root
-        else:
-            for file_name in fs.ls(path):
-                if not match_masks(file_name, self.masks):
-                    continue
-
-                # ensure the file name has the full fsspec protocol path
-                if file_name.startswith(fs.protocol):
-                    yield file_name
-                else:
-                    if is_local:
-                        abs_path = os.path.join(path, file_name)
-                    else:
-                        abs_path = posixpath.join(path, file_name)
-
-                    if self.root.startswith(fs.protocol):
-                        yield fs.protocol + "://" + abs_path
+        for root in self.datapipe:
+            fs, path = fsspec.core.url_to_fs(root)
+
+            if isinstance(fs.protocol, str):
+                protocol_list = [fs.protocol]
+            else:
+                protocol_list = fs.protocol
+
+            is_local = fs.protocol == "file" or not any(root.startswith(protocol) for protocol in fs.protocol)
+            if fs.isfile(path):
+                yield root
+            else:
+                for file_name in fs.ls(path):
+                    if not match_masks(file_name, self.masks):
+                        continue
+
+                    # ensure the file name has the full fsspec protocol path
+                    if any(file_name.startswith(protocol) for protocol in protocol_list):
+                        yield file_name
                     else:
-                        yield abs_path
+                        if is_local:
+                            abs_path = os.path.join(path, file_name)
+                        else:
+                            abs_path = posixpath.join(path, file_name)
+
+                        starts_with = False
+                        for protocol in protocol_list:
+                            if root.startswith(protocol):
+                                starts_with = True
+                                yield protocol + "://" + abs_path
```

Comment:
Let's add a `break` when the root path starts with the protocol the first time. Otherwise, there can be multiple paths yielded.